### PR TITLE
SNOW-767903: Resolve the race condition of deleting temp file when running PUT

### LINKED ISSUE #1490: SNOW-767903: "OSError: [WinError 145] The directory is not empty" when PUTing a temporary stage
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

Python 3.7.12

2. What operating system and processor architecture are you using?

Windows-2012ServerR2-6.3.9600-SP0

3. What are the component versions in the environment (`pip freeze`)?

[I've removed some internal modules from this output.]

```
absl-py==1.2.0
aenum==3.1.11
asn1crypto==1.5.1
astunparse==1.6.3
asyncpg==0.25.0
atomicwrites==1.4.0
attrs==21.4.0
autoflake==1.4
backoff==2.2.1
backports.zoneinfo==0.2.1
black==23.1.0
boto3==1.21.44
boto3-stubs==1.21.44
botocore==1.24.44
botocore-stubs==1.24.44
bump2version==1.0.1
bumpversion==0.6.0
cachetools==5.2.0
ceODBC==2.0.1
certifi==2021.10.8
cffi==1.15.0
chardet==4.0.0
charset-normalizer==2.0.12
click==8.1.2
colorama==0.4.4
convertdate==2.4.0
cryptography==36.0.2
cycler==0.11.0
dataclasses-json==0.5.7
dateparser==1.1.1
decorator==5.1.1
energyquantified==0.8
et-xmlfile==1.1.0
exceptiongroup==1.0.1
filelock==3.8.0
flatbuffers==23.1.4
fonttools==4.32.0
gast==0.4.0
google-auth==2.9.1
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
greenlet==1.1.2
grpcio==1.47.0
h5py==3.7.0
hijri-converter==2.2.3
holidays==0.21.13
hypothesis==6.70.0
idna==2.10
importlib-metadata==4.11.3
iniconfig==1.1.1
isort==5.10.1
Jinja2==3.1.2
jmespath==1.0.0
joblib==1.1.0
keras==2.11.0
Keras-Preprocessing==1.1.2
kiwisolver==1.4.2
korean-lunar-calendar==0.2.1
libclang==14.0.1
libcst==0.4.7
Markdown==3.4.1
MarkupSafe==2.1.1
marshmallow==3.15.0
marshmallow-enum==1.5.1
matplotlib==3.5.3
mypy==1.1.1
mypy-boto3-cloudformation==1.21.34
mypy-boto3-cloudwatch==1.21.41
mypy-boto3-dynamodb==1.21.34
mypy-boto3-ec2==1.21.41
mypy-boto3-ecs==1.21.34
mypy-boto3-lambda==1.21.35
mypy-boto3-logs==1.21.34
mypy-boto3-rds==1.21.44
mypy-boto3-s3==1.21.34
mypy-boto3-sqs==1.21.34
mypy-extensions==1.0.0
networkx==2.5
numpy==1.21.6
oauthlib==3.2.0
openpyxl==3.1.0
openpyxl-stubs==0.1.25
opt-einsum==3.3.0
oscrypto==1.3.0
packaging==23.0
pandas==1.3.5
pandas-stubs==1.2.0.62
pantab==1.1.1
pathspec==0.9.0
patsy==0.5.2
Pillow==9.1.0
platformdirs==2.5.2
pluggy==1.0.0
protobuf==3.19.4
psycopg2==2.9.3
py==1.11.0
pyappcache==0.9
pyarrow==10.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.21
pycryptodomex==3.15.0
pyflakes==2.4.0
PyJWT==2.5.0
PyMeeus==0.5.11
pymsteams==0.2.1
pyodbc==4.0.32
pyOpenSSL==22.0.0
pyparsing==3.0.8
pytest==7.1.1
pytest-stub==1.1.0
python-dateutil==2.8.2
pytz==2022.1
pytz-deprecation-shim==0.1.0.post0
pyupgrade==3.0.0
PyYAML==6.0
regex==2022.3.2
requests==2.28.2
requests-oauthlib==1.3.1
rsa==4.9
s3transfer==0.5.2
scikit-learn==1.0.2
scipy==1.7.3
sentry-sdk==1.5.10
shed==2023.3.1
six==1.16.0
sklearn==0.0
snowflake-connector-python==3.0.0
snowflake-sqlalchemy==1.4.6
sortedcontainers==2.4.0
SQLAlchemy==1.4.35
sqlparse==0.4.2
statsmodels==0.13.5
tableauhyperapi==0.0.14567
tensorboard==2.11.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow==2.11.0
tensorflow-estimator==2.11.0
tensorflow-intel==2.11.0
tensorflow-io-gcs-filesystem==0.26.0
termcolor==1.1.0
threadpoolctl==3.1.0
timer==0.2.2
tokenize-rt==4.2.1
tomli==2.0.1
typed-ast==1.5.3
types-docutils==0.19.1.1
types-psycopg2==2.9.21.8
types-pytz==2022.7.1.2
types-requests==2.28.11.15
types-setuptools==67.6.0.5
types-urllib3==1.26.13
typing-inspect==0.7.1
typing_extensions==4.4.0
tzdata==2022.1
tzlocal==4.2
urllib3==1.26.13
Werkzeug==2.2.0
wrapt==1.14.1
xgboost==1.6.2
zipp==3.8.0
```

4. What did you do?

Infrequently, when calling PUT to upload a file into a stage, we get an OSError.  Here is a truncated traceback:

```python
INFO:snowflake.connector.cursor:query: [CREATE TEMPORARY STAGE tmp_20230324_p0gychps7ze7pg;]
INFO:snowflake.connector.cursor:query execution done
INFO:snowflake.connector.cursor:Number of results in first chunk: 1
INFO:snowflake.connector.cursor:query: [PUT file://C:\Users\WindowsUser\AppData\Local\Temp\2\tmpae78m5f_ @tmp_20230324_p0gy...]
INFO:snowflake.connector.cursor:query execution done
INFO:snowflake.connector.cursor:query: [ROLLBACK]
INFO:snowflake.connector.cursor:query execution done
INFO:snowflake.connector.cursor:Number of results in first chunk: 1
INFO:snowflake.connector.cursor:query: [ROLLBACK]
INFO:snowflake.connector.cursor:query execution done
INFO:snowflake.connector.cursor:Number of results in first chunk: 1
Traceback (most recent call last):
[[ lines removed ]]
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\project_module_dir\code.py", line 140, in _snowflake_insert_copy
    cur.execute(f"PUT file://{temp_file_path} @{temp_stage_name};")
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\cursor.py", line 794, in execute
    sf_file_transfer_agent.execute()
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\file_transfer_agent.py", line 404, in execute
    self.transfer(self._file_metadata)
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\file_transfer_agent.py", line 622, in transfer
    raise exception_caught_in_callback
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\file_transfer_agent.py", line 578, in function_and_callback_wrapper
    _callback(*result, file_meta)
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\file_transfer_agent.py", line 521, in transfer_done_cb
    done_client.delete_client_data()
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\site-packages\snowflake\connector\storage_client.py", line 439, in delete_client_data
    shutil.rmtree(self.tmp_dir)
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\shutil.py", line 516, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\shutil.py", line 404, in _rmtree_unsafe
    onerror(os.rmdir, path, sys.exc_info())
  File "C:\Users\WindowsUser\.conda\envs\project_dir\lib\shutil.py", line 402, in _rmtree_unsafe
    os.rmdir(path)
OSError: [WinError 145] The directory is not empty: 'C:\\Users\\WindowsUser\\AppData\\Local\\Temp\\2\\tmpjb6qe5gd'
```

I think this is happening for a windows-specific reason - *windows doesn't allow files with open file handles to be deleted*.  But I'm not sure.  I think it's possible that something is still writing to the tmpdir (perhaps flushing a buffered file?) while `shutil.rmtree` is running.  I estimate this error occurs about 1-2% of the time.

That means I'm not able to reproduce this reliably, sorry.

5. What did you expect to see?

I expect the PUT to be completed successfully.

6. Can you set logging to DEBUG and collect the logs?

I've included some logging output above.  If more is needed I can share that over email.

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1490

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   We shouldn't delete files in `__del__` without lock. See details in https://stackoverflow.com/a/66340298

### CODE REVIEW

File: src/snowflake/connector/storage_client.py [hunk 1]
```diff
@@ -445,9 +445,10 @@ def _abort_multipart_upload(self) -> None:
     def delete_client_data(self) -> None:
         """Deletes the tmp_dir and closes the source stream belonging to this client.
         This function is idempotent."""
-        if os.path.exists(self.tmp_dir):
-            logger.debug(f"cleaning up tmp dir: {self.tmp_dir}")
-            shutil.rmtree(self.tmp_dir)
+        with self.lock:
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/storage_client.py:448]: The stackoverflow answer claims that `An object deconstructor may not always have a predictable behaviour` but for CPython, PEP-442 mandates that an object’s finalizer should always called exactly once. The stackoverflow answer states that their race condition comes from using `multiprocessing` where two processes try to delete the same directory, but that shouldn't apply to us. I'm curious as to where the race condition is coming from here.

Comment (bml0aGlu) [reply src/snowflake/connector/storage_client.py:448]: possibly from this function: _upload_files_in_parallel? In my case, I am using older snowflake connector package and I have the same issue.

File: src/snowflake/connector/storage_client.py [hunk 2]
```diff
@@ -445,9 +445,10 @@ def _abort_multipart_upload(self) -> None:
...
+            if os.path.exists(self.tmp_dir):
+                logger.debug(f"cleaning up tmp dir: {self.tmp_dir}")
+                shutil.rmtree(self.tmp_dir)
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/storage_client.py:451]: I think it'd be acceptable here to just catch all exceptions coming out of `shutil.rmtree` and ignore them. And so if 2 threads are cleaning up simultaneously then if 1 of them is already cleaning up the other one could just give up on doing the same thing.

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**